### PR TITLE
Fix for agent hang when process cannot be killed

### PR DESCRIPTION
**Pull request by merlynomsft**
This PR fixes an issue where the agent will hang and fail to upload logs if the process executed cannot be killed during cancellation.   The pipeline hangs until timeoutInMinutes  + cancelTimeoutInMinutes

The log fails to upload and a blank log is shown in the log viewer UX:
image

**Analysis**
Killing process during cancellation is 'best effort' and may fail if, for some reason, the process cannot be killed.  In that case, the agent waits forever in a loop for the process to exit.  Fix is to exit the loop after attempting to kill the process.

**Testing**
This issue is not easy to reproduce, because it requires the agent to launch a process that cannot be killed
I simulated this condition by modifying ProcessInvoker to skip killing the process

**Knob / Enabling **
The behavior is disabled by default.  This change is enabled by setting knob: VSTSAGENT_CONTINUE_AFTER_CANCEL_PROCESSTREEKILL_ATTEMPT

VSO 2009025

### CODE REVIEW

File: src/Agent.Sdk/ProcessInvoker.cs [hunk 1]
```diff
@@ -415,8 +443,11 @@ private void ProcessOutput()
             }
         }
 
-        private async Task CancelAndKillProcessTree(bool killProcessOnCancel)
+        internal protected virtual async Task CancelAndKillProcessTree(bool killProcessOnCancel)
         {
+            Trace.Info("skipping CancelAndKillProcessTree");
```
Comment (merlynomsft) [inline src/Agent.Sdk/ProcessInvoker.cs]: [test code] Pushed this in by mistake - should not be present

**Review by DenisRumyantsev — APPROVED** [review]
@merlynomsft thank you so much for your work!
I artificially reproduced the issue to test these changes.
The agent is working properly and the issue is resolved.

File: src/Agent.Sdk/Knob/AgentKnobs.cs [hunk 2]
```diff
@@ -370,6 +370,14 @@ public class AgentKnobs
...
 
+        public const string ContinueAfterCancelProcessTreeKillAttemptVariableName = "VSTSAGENT_CONTINUE_AFTER_CANCEL_PROCESSTREEKILL_ATTEMPT";
+
+        public static readonly Knob ContinueAfterCancelProcessTreeKillAttempt = new Knob(
+            nameof(ContinueAfterCancelProcessTreeKillAttempt),
+            "If true, continue cancellation after attempt to KillProcessTree",
+            new EnvironmentKnobSource(ContinueAfterCancelProcessTreeKillAttemptVariableName),
+            new BuiltInDefaultKnobSource("true"));
+
         public static readonly Knob UseNode = new Knob(
```
[the remark this answers is not in the document]
Comment (merlynomsft) [reply src/Agent.Sdk/Knob/AgentKnobs.cs:382]: An update -- with the latest push, the behavior is now gated on a RuntimeKnobSource, off by default.  To enable, set VSTSAGENT_CONTINUE_AFTER_CANCEL_PROCESSTREEKILL_ATTEMPT as a variable in the run.  Intent is to use a server-side feature flag to gate/rollout the new behavior.

**Review by LiliaSabitova — APPROVED** [review]

**Review by kirill-ivlev — APPROVED** [review]

**Review by vmapetr — APPROVED** [review]

**Review by wojgasior — APPROVED** [review]

**Comment by merlynomsft** [comment]
Please note currently this behavior is opt-in.  To do so, set the following environment variable on the agent:

VSTSAGENT_CONTINUE_AFTER_CANCEL_PROCESSTREEKILL_ATTEMPT: true

**Comment by merlynomsft** [comment]
Please note currently this change is opt-in.  To do so, set the following environment variable on the agent:

VSTSAGENT_CONTINUE_AFTER_CANCEL_PROCESSTREEKILL_ATTEMPT: true

**Pull request merged**